SonarQube.Client/Models/ServerSentEvents/SSEStreamWriter.cs

```diff
@@ -35,39 +36,60 @@ public interface ISSEStreamWriter : IDisposable
         Task BeginListening();
     }
 
+    /// <summary>
+    /// Aggregates stream lines into events.
+    /// Code on the java side: https://github.com/SonarSource/sonarlint-core/blob/171ca4d75c24033e115a81bd7481427cd1f39f4c/server-api/src/main/java/org/sonarsource/sonarlint/core/serverapi/stream/EventBuffer.java
+    /// </summary>
     internal class SSEStreamWriter : ISSEStreamWriter
     {
-        private readonly Stream serverStream;
+        private readonly StreamReader networkStreamReader;
         private readonly ChannelWriter<ISqServerEvent> sqEventsChannel;
         private readonly CancellationToken cancellationToken;
         private readonly ISqServerSentEventParser sqServerSentEventParser;
 
-        public SSEStreamWriter(Stream serverStream,
+        public SSEStreamWriter(StreamReader networkStreamReader,
             ChannelWriter<ISqServerEvent> sqEventsChannel,
             CancellationToken cancellationToken)
-            : this(serverStream, sqEventsChannel, cancellationToken, new SqServerSentEventParser())
+            : this(networkStreamReader, sqEventsChannel, cancellationToken, new SqServerSentEventParser())
         {
         }
 
-        internal SSEStreamWriter(Stream serverStream,
+        internal SSEStreamWriter(StreamReader networkStreamReader,
             ChannelWriter<ISqServerEvent> sqEventsChannel,
             CancellationToken cancellationToken,
             ISqServerSentEventParser sqServerSentEventParser)
         {
-            this.serverStream = serverStream;
+            this.networkStreamReader = networkStreamReader;
             this.cancellationToken = cancellationToken;
             this.sqEventsChannel = sqEventsChannel;
             this.sqServerSentEventParser = sqServerSentEventParser;
         }
 
-        public Task BeginListening()
+        public async Task BeginListening()
         {
-            // todo: next PR
-            throw new NotImplementedException();
+            var eventLines = new List<string>();
+
+            while (!networkStreamReader.EndOfStream && !cancellationToken.IsCancellationRequested)
+            {
+                var line = await networkStreamReader.ReadLineAsync();
+                var isEventEnd = string.IsNullOrEmpty(line);
+
+                if (isEventEnd)
+                {
+                    var parsedEvent = sqServerSentEventParser.Parse(eventLines);
+                    await sqEventsChannel.WriteAsync(parsedEvent, cancellationToken);
+                    eventLines.Clear();
+                }
+                else
+                {
+                    eventLines.Add(line);
+                }
+            }
```

Comment:
What will happen in the event of an exception? (e.g. user loses internet connection briefly)

What is responsible for notification and reconnecting?

Reply:
You're right, I missed this. Opened separate tasks.